.github/actions/setup-and-login/action.yml

```diff
@@ -0,0 +1,67 @@
+name: Setup Node, Env and login to GHCR
+
+inputs:
+  install-packages:
+    required: false
+    default: 'false'
+  build-and-run-server:
+    required: false
+    default: 'false'
+  ghcr-username:
+    required: false
+    default: ''
+  ghcr-password:
+    required: false
+    default: ''
+
+outputs:
+  CI_SHA:
+    description: "Short SHA of current commit"
+    value: ${{ steps.ci-env.outputs.CI_SHA }}
+  CI_BRANCH:
+    description: "Current branch"
+    value: ${{ steps.ci-env.outputs.CI_BRANCH }}
+
+runs:
+  using: "composite"
+  steps:
+    - name: Login to GHCR (GitHub Packages)
+      if: inputs.ghcr-username != '' && inputs.ghcr-password != ''
+      uses: docker/login-action@v3
+      with:
+        registry: ghcr.io
+        username: ${{ inputs.ghcr-username }}
+        password: ${{ inputs.ghcr-password }}
+    - id: ci-env
+      name: Setup Environment
+      shell: bash
+      run: |
+        if [ "${{github.event_name}}" = "pull_request" ];
+        then
+          long_sha=${{ github.event.pull_request.head.sha }}
+          echo "CI_BRANCH=${{ github.head_ref }}" >> $GITHUB_OUTPUT
+        else
+          long_sha=${GITHUB_SHA}
+          echo "CI_BRANCH=${{ github.ref_name }}" >> $GITHUB_OUTPUT
+        fi
+        echo "CI_SHA=${long_sha:0:7}" >> $GITHUB_OUTPUT
+
+    - uses: actions/setup-node@v4
+      with:
+        node-version: 20
+
+    - name: Install packages
+      # we install in the build script anyway so skip if build-and-run-server
+      if: inputs.install-packages == 'true' && inputs.build-and-run-server != 'true'
+      shell: bash
+      run: |
+        npm i --prefix=app/server
+        npm i --prefix=app/static
+
+    - name: Build and run server
+      if: inputs.build-and-run-server == 'true'
+      shell: bash
+      run: |
+        ./scripts/build.sh
+        ./scripts/run-dependencies.sh
+        npm run serve --prefix=app/server &
```

Comment:
Again, this feels a bit like it's outside the remit of this action.Is this action really a single bucket in which to put everything that might be shared between jobs so that there is a one liner for set-up in the job? But then that one liner ends up with a gruesome string of parameters potentially (or confusing lack of parameters, relying on defaults)? It's quite subjective but personally I feel like it would be a bit nicer to split out the actions into single-responsibility parts, and call the ones you need from each job. Node with install packages as an option feels like one action for example. 

I suppose the advantage of doing it this way is that the mega-action knows what order things need to happen in, and how the parameters might impact each other. But that feels a little more brittle too, compared to each job just picking the bits it needs.

Reply:
yhhh i was treating this as a utils file basically, but yh probably cleaner to have this as multiple actions, i have split into 3 actions, one for ghcr and ci env login, one for npm packages and the last for running the entire app

the reasom i was hesitant is because there is so much overhead in creating a new action that even if the code feels cleaner, it is messier to navigate because the code isnt as local, just to parse through `test.yml` you have to navigate to 3 other files now instead of 1, but youre right the code itself is cleaner and more logical, i just wish there was a less costly way of writing reusable gh actions code

this is also why i partly hate navigating kotlin code (obviously not gonna miss the opportunity to slate spring boot), you have 3-4 different files all doing a little bit and passing the parcel along, logic spread across files is much harder at least for me to follow

btw im just ranting, im happy to keep this as 3 different actions XD this is what was going through my head

Reply:
Yeah, I know what you mean. At least here the actual action implementation only needs to happen in one place per action!